### PR TITLE
implemented validation string operations in java target.

### PR DESCRIPTION
**Pull request by vadlis**
In fact - copied from security.php

### CODE REVIEW

**Review by bvaschenko — APPROVED** [review]
LGTM

File: lib/password.flow [hunk 1]
```diff
@@ -5,6 +5,17 @@ export {
 		password : string, 
 		hash : string 
 	) -> bool = FlowPassword.validateHash;
+
+	// < 0: means no expiration
+	// 0: use default expiration time (84600 sec -> 1 day)
+	// > 0: use specified value 
+	native getPasswordValidationStringExp : (exp: int) -> string = FlowPassword.getPasswordValidationString;
```
Comment (sfandersen) [inline lib/password.flow:12]: Instead of "exp" it is better to call it "expirationS". (I assume it is in seconds).

File: lib/password.flow [hunk 2]
```diff
@@ -5,6 +5,17 @@ export {
...
+	// get validation string without expiration (create_reset_password_validation_string(null))
+	getPasswordValidationString : () -> string;
+
+	native getValidationStringExpiration : (vs : string) -> int = FlowPassword.getValidationStringExpiration;
+	native checkValidationStringExpirationExp : (vs : string, exp: int) -> bool = FlowPassword.checkValidationStringExpiration;
```
Comment (sfandersen) [inline lib/password.flow:17]: It would seem to me like the checkValidationStringExpirationExp does not need to be an native?

Comment (vadlis) [reply lib/password.flow:17]: agree

File: platforms/java/com/area9innovation/flow/FlowPassword.java [hunk 3]
```diff
@@ -102,4 +106,68 @@ public static Boolean validateHash(String password, String hash) {
...
+
+		try {
+			JSONObject jo = new JSONObject();
+			jo.put("key", Base64.getEncoder().encodeToString(bytes));
+			if (expLocal != null) {
+				long unixTime = System.currentTimeMillis() / 1000L;
+				jo.put("exp", unixTime + expLocal);
+			};
+
+			validation = strtr(Base64.getEncoder().encodeToString(jo.toString().getBytes()), "+/=", "*_-");
```
Comment (sfandersen) [inline platforms/java/com/area9innovation/flow/FlowPassword.java:126]: I think you can just do "Base64.getUrlDecoder", that way you do not need the strtr function (If you need it, rename it to something like replaceCharaters

Comment (vadlis) [reply platforms/java/com/area9innovation/flow/FlowPassword.java]: I think we need this to produce the same string as security.php do. So string from java & php are the same and can be processed in both targets

**Review by sfandersen — APPROVED** [review]
I had a few comments. Otherwise it looks good.

File: platforms/java/com/area9innovation/flow/FlowPassword.java (hunk 3 above)
Comment (sfandersen) [reply platforms/java/com/area9innovation/flow/FlowPassword.java]: Right. Agree, the * and _ are not normal base64url encode.

**Pull request merged**